Use phantom data trackers to auto drop

cholcombe973 opened the pull request:
This change adds automatic closing of ceph connections where necessary.  Unfortunately it required reworking a huge amount of the library.  The most important pieces are the impl Drop sections in src/ceph.rs

The basics here are I wrapped rados_t with Rados and rados_ioctx_t with IoCtx.  I made the inner fields public because my ceph-rbd library needs to get access to those pointers.  There's probably a better way to do that? 

I checked that this worked with a small [rayon program](https://gist.github.com/cholcombe973/8ea370c8ca37a50be17e8a900057cf0d) that runs rbd stat in parallel.  This would consistently crash until I got the Rbd, ioctx and ceph cluster connection dropping in the right order.  

It now runs fine with 10 threads and valgrind shows it's clean:
```==1423173== 
==1423173== HEAP SUMMARY:
==1423173==     in use at exit: 0 bytes in 0 blocks
==1423173==   total heap usage: 408,707 allocs, 408,707 frees, 19,222,091 bytes allocated
==1423173== 
==1423173== All heap blocks were freed -- no leaks are possible
==1423173== 
==1423173== For counts of detected and suppressed errors, rerun with: -v
==1423173== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

I may have messed up your ceph_client.rs and cmd.rs @ChrisMacNaughton  so take a close look at those

cholcombe973 committed:
Use phantom data trackers to auto drop

This change adds automatic closing of ceph connections where
necessary.  Unfortunately it required reworking a huge amount
of the library.  The most important pieces are the impl Drop
sections in src/ceph.rs

cholcombe973 committed:
Fix up tests

cholcombe973 committed:
Remove drop from CephClient

That isn't needed anymore

cholcombe973 commented:
Related PR for ceph-rbd: https://github.com/cholcombe973/ceph-rbd/pull/5

ChrisMacNaughton commented on examples/ceph.rs:
```diff
@@ -17,35 +17,37 @@
 use ceph::ceph as ceph_helpers;
 #[cfg(target_os = "linux")]
 use ceph::rados;
+use ceph::JsonData;
```
Can we try to do things like rust-fmt separately; it massively increases the difficulty of reading through the changes on this review

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -65,53 +65,44 @@ pub(crate) fn get_error(n: c_int) -> RadosResult<String> {
-        data_len: le_u32 >>
-        data: take!(data_len) >>
-        (TmapOperation::Create{
+        char!(CEPH_OSD_TMAP_CREATE) >> key_name_len: le_u32 >> key_name: take_str!(key_name_len) >> data_len: le_u32
```
This is bad auto-formatting, should be reverted

ChrisMacNaughton replied:
That may be easier to do as an upgrade to `nom` rather than convincing rust-fmt to not f*** up this file

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-        Ok(String::from_utf8_lossy(&buffer).into_owned())
-    }
+/// Owns a ioctx handle
+pub struct IoCtx {
```
Does the `IoCtx` struct itself need to be public? I understand the `Rados` needing to be `pub` but I wonder where we can restrict privacy to needed public types

cholcombe973 replied:
Good question.  For the ceph-rbd library it needs to access IoCtx for pretty much everything.  It's making calls down to librbd which need the pointer passed over to it.  I keep waffling back and forth between having an access method, folding ceph-rbd into ceph or leaving the inner field public.  I think leaving the inner field public is dangerous.  The access method might be ok if we can have rust take care of what we hand out I think.

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-    }
+/// Owns a ioctx handle
+pub struct IoCtx {
+    pub ioctx: rados_ioctx_t,
```
rather than making the `ioctx` itself pub, maybe the containing struct could have an accessor that passes out an `&mut rados_ioctx_t` or whatever format is needed?

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-    }
+/// Owns a rados handle
+pub struct Rados {
+    pub rados: rados_t,
```
same thought as above regarding having an accessor rather than actually making this a public field

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-        let ret_code = rados_ioctx_pool_required_alignment(ctx);
-        return Ok(ret_code);
+    /// Set the value of a configuration option
+    /// NOTE: the rados_t parameter must not be connected to ceph yet when this
```
Given this imperative, it seems like something we should check for in code

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
+            // Ceph doesn't return how many bytes were written
+            buffer.set_len(5120);
+            // We need to search for the first NUL byte
+            let num_bytes = buffer.iter().position(|x| x == &0u8);
```
I wonder how this performs compared to something like using [memchr](https://github.com/BurntSushi/rust-memchr)

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
+    /// requests on it, but you
+    /// should not use an io context again after calling this function on it.
+    /// This does not guarantee any asynchronous writes have completed. You must
+    /// call rados_aio_flush()
```
Seems like something that we can do for the caller unless some flag is set on the struct maybe?

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-    }
-    let object_name_str = try!(CString::new(object_name));
+    /// The object is filled with the provided data. If the object exists, it is
+    /// atomically
```
extra newline

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
+    /// If the underlying filesystem on the OSD supports it, this will be a
+    /// copy-on-write clone.
+    /// The src and dest objects must be in the same pg. To ensure this, the io
+    /// context should
```
Another extra newline

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-    let object_name_str = try!(CString::new(object_name));
-    let class_name_str = try!(CString::new(class_name));
-    let method_name_str = try!(CString::new(method_name));
+    // pub fn rados_object_tmap_put(ctx: rados_ioctx_t, object_name: &str) ->
```
unnecessary comments?

cholcombe973 replied:
I think that was a function I was planning on implementing but never did.  I'm deleting it now.

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -277,1337 +264,1422 @@ pub struct RadosVersion {
-
-    match buffer {
-        Some(buf) => unsafe {
-            let ret_code = rado
```
unnecessary comments?

cholcombe973 replied:
I'll delete these also.

ChrisMacNaughton commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ceph"
-version = "0.5.0"
+version = "0.6.0"
```
This shouldn't be changed until we're ready to release a new version, at which point it should be a major version change

ChrisMacNaughton reviewed (changes requested):
Lots of inline comments

cholcombe973 committed:
Remove formatting

cholcombe973 committed:
Address review comments

ChrisMacNaughton commented on src/ceph.rs:
```diff
@@ -382,6 +382,10 @@ pub fn connect_to_ceph<'a>(user_id: &str, config_file: &str) -> RadosResult<Rado
 
 impl Rados {
+    pub fn inner(&self) -> rados_t {
+        self.rados
```
do we really want ot pass ownership of the pointer out?

cholcombe973 committed:
Undo rustfmt

cholcombe973 committed:
Fix up example again

cholcombe973 committed:
Undo rustfmt

cholcombe973 committed:
Lend pointers

If the user requires a pointer the inner() function will give it
to them without losing ownership of it *I think*

cholcombe973 committed:
Fix tests

cholcombe973 commented:
valgrind still returns clean :)

cholcombe973 commented:
Since this works should we remove the disconnect_from_ceph and destroy_rados_ioctx functions?

ChrisMacNaughton commented:
Only if we're assuming that the user will never use the pointers directly; it may make more sense to look at splitting the FFI bits into a sub-crate like `ceph-sys`, and keeping only the higher level abstractions in the primary crate

cholcombe973 commented:
Yeah I'm for `ceph-sys`.  I can break that out in another PR

ChrisMacNaughton commented:
I'd love for some time to be spent identifying the travis failure, but the only reference for it that I can find is from a cargo bug a couple of years ago regarding badly packaged crate archives
